File: sansa-query/sansa-query-spark/src/main/scala/net/sansa_stack/query/spark/binding/engine/ExecutionDispatch.scala
package net.sansa_stack.query.spark.binding.engine

import java.util

import org.apache.jena.sparql.algebra.op._
import org.apache.jena.sparql.algebra.{Op, OpVisitor}
import org.apache.jena.sparql.engine.binding.Binding
import org.apache.spark.rdd.RDD

class ExecutionDispatch(opExecutor: OpExecutor)
  extends OpVisitor
{
  val stack = new util.ArrayDeque[RDD[Binding]]()

  def exec(op: Op, input: RDD[Binding]): RDD[Binding] = {
    stack.push(input)
    val x = stack.size
    op.visit(this)
    val y = stack.size
    if (x != y) throw new RuntimeException("Possible stack misalignment")
    val result = stack.pop()
    result
  }
  override def visit(opBGP: OpBGP): Unit = ???

  override def visit(quadPattern: OpQuadPattern): Unit = ???

  override def visit(quadBlock: OpQuadBlock): Unit = ???

  override def visit(opTriple: OpTriple): Unit = ???

  override def visit(opQuad: OpQuad): Unit = ???

  override def visit(opPath: OpPath): Unit = ???

  override def visit(opFind: OpFind): Unit = ???

  override def visit(opTable: OpTable): Unit = ???

  override def visit(opNull: OpNull): Unit = ???

  override def visit(opProc: OpProcedure): Unit = ???

  override def visit(opPropFunc: OpPropFunc): Unit = ???

  override def visit(opFilter: OpFilter): Unit =
    stack.push(opExecutor.execute(opFilter, stack.pop()))

  override def visit(opGraph: OpGraph): Unit = ???

  override def visit(opService: OpService): Unit =
    stack.push(opExecutor.execute(opService, stack.pop()))

  override def visit(dsNames: OpDatasetNames): Unit = ???

  override def visit(opLabel: OpLabel): Unit = ???

  override def visit(opAssign: OpAssign): Unit = ???

  override def visit(opExtend: OpExtend): Unit =
    stack.push(opExecutor.execute(opExtend, stack.pop()))

  override def visit(opJoin: OpJoin): Unit = ???

  override def visit(opLeftJoin: OpLeftJoin): Unit = ???

  override def visit(opUnion: OpUnion): Unit =
    stack.push(opExecutor.execute(opUnion, stack.pop()))

  override def visit(opDiff: OpDiff): Unit = ???

  override def visit(opMinus: OpMinus): Unit = ???

  override def visit(opCondition: OpConditional): Unit = ???

  override def visit(opSequence: OpSequence): Unit = ???

  override def visit(opDisjunction: OpDisjunction): Unit = ???

  override def visit(opList: OpList): Unit = ???

  override def visit(opOrder: OpOrder): Unit =
    stack.push(opExecutor.execute(opOrder, stack.pop))

  override def visit(opProject: OpProject): Unit =
    stack.push(opExecutor.execute(opProject, stack.pop))

  override def visit(opReduced: OpReduced): Unit =
    stack.push(opExecutor.execute(opReduced, stack.pop))

  override def visit(opDistinct: OpDistinct): Unit =
    stack.push(opExecutor.execute(opDistinct, stack.pop))


  override def visit(opSlice: OpSlice): Unit =
    stack.push(opExecutor.execute(opSlice, stack.pop))


  override def visit(opGroup: OpGroup): Unit =
    stack.push(opExecutor.execute(opGroup, stack.pop))


  override def visit(opTop: OpTopN): Unit = ???
}

File: sansa-query/sansa-query-spark/src/main/java/net/sansa_stack/query/spark/engine/ExecutionDispatch.java
package net.sansa_stack.query.spark.engine;

import org.apache.jena.sparql.algebra.Op;
import org.apache.jena.sparql.algebra.OpVisitor;
import org.apache.jena.sparql.algebra.op.*;
import org.apache.jena.sparql.engine.binding.Binding;
import org.apache.spark.api.java.JavaRDD;

import java.util.ArrayDeque;
import java.util.Deque;

public class ExecutionDispatch
        implements OpVisitor {
    protected OpExecutor opExecutor;

    public ExecutionDispatch(OpExecutor opExecutor) {
        this.opExecutor = opExecutor;
    }

    protected Deque<JavaRDD<Binding>> stack = new ArrayDeque<>();

    public JavaRDD<Binding> exec(Op op, JavaRDD<Binding> input) {
        stack.push(input);
        int x = stack.size();
        op.visit(this);
        int y = stack.size();
        if (x != y) throw new RuntimeException("Possible stack misalignment");
        JavaRDD<Binding> result = stack.pop();
        return result;
    }

    @Override
    public void visit(OpBGP opBGP) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void visit(OpQuadPattern quadPattern) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void visit(OpQuadBlock quadBlock) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void visit(OpTriple opTriple) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void visit(OpQuad opQuad) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void visit(OpPath opPath) {
        throw new UnsupportedOperationException();
    }

    // @Override public void visit(opFind: OpFind){ throw new UnsupportedOperationException(); }

    @Override
    public void visit(OpTable opTable) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void visit(OpNull opNull) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void visit(OpProcedure opProc) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void visit(OpPropFunc opPropFunc) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void visit(OpFilter opFilter) {
        stack.push(opExecutor.execute(opFilter, stack.pop()));
    }

    @Override
    public void visit(OpGraph opGraph) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void visit(OpService opService) {
        stack.push(opExecutor.execute(opService, stack.pop()));
    }

    @Override
    public void visit(OpDatasetNames dsNames) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void visit(OpLabel opLabel) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void visit(OpAssign opAssign) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void visit(OpExtend opExtend) { stack.push(opExecutor.execute(opExtend, stack.pop())); }

    @Override
    public void visit(OpJoin opJoin) { stack.push(opExecutor.execute(opJoin, stack.pop())); }

    @Override
    public void visit(OpLeftJoin opLeftJoin) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void visit(OpUnion opUnion) {
        stack.push(opExecutor.execute(opUnion, stack.pop()));
    }

    @Override
    public void visit(OpDiff opDiff) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void visit(OpMinus opMinus) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void visit(OpConditional opCondition) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void visit(OpSequence opSequence) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void visit(OpDisjunction opDisjunction) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void visit(OpList opList) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void visit(OpOrder opOrder) {
        stack.push(opExecutor.execute(opOrder, stack.pop()));
    }

    @Override
    public void visit(OpProject opProject) {
        stack.push(opExecutor.execute(opProject, stack.pop()));
    }

    @Override
    public void visit(OpReduced opReduced) {
        stack.push(opExecutor.execute(opReduced, stack.pop()));
    }

    @Override
    public void visit(OpDistinct opDistinct) {
        stack.push(opExecutor.execute(opDistinct, stack.pop()));
    }

    @Override
    public void visit(OpSlice opSlice) {
        stack.push(opExecutor.execute(opSlice, stack.pop()));
    }


    @Override
    public void visit(OpGroup opGroup) {
        stack.push(opExecutor.execute(opGroup, stack.pop()));
    }

    @Override
    public void visit(OpTopN opTop) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void visit(OpLateral opLateral) { stack.push(opExecutor.execute(opLateral, stack.pop())); }
}
